### PR TITLE
fix(forge) hotfix - filter tests

### PR DESCRIPTION
**Pull request by iFrostizz**
Hotfix for https://github.com/foundry-rs/foundry/pull/5753
Closes https://github.com/foundry-rs/foundry/issues/5563

I have introduced a bug that didn't filtered the tests correctly using the construct.
Fixes and ~slight refactor so that won't repeat cause we don't depend on 2 instance of the same code.
Also we don't need to turn off debug traces, what I initially thought was a big performance penalty was just proptest shrinking while it wasn't ran in `--release` mode

### CODE REVIEW

**Review by Evalir — APPROVED** [review]
lgtm, are we sure this fixes filtering? is anything else broken right now?

**Review by Evalir — CHANGES_REQUESTED** [review]
needs fixing on tests

**Comment by iFrostizz** [comment]
Yeah @Evalir , `forge debug` is broken with "Invalid hex calldata" but I'm wondering if we shouldn't just nuke it at this point, c.f. https://github.com/foundry-rs/foundry/pull/4793#discussion_r1174305656
tldr it serves the same purpose as `forge script --debug` and is just redundant
You can see that it's `forge script` with less stuff, it's even calling `run_script` soooo https://github.com/foundry-rs/foundry/blob/master/crates/forge/bin/cmd/debug.rs#L43-L56

**Comment by Evalir** [comment]
@iFrostizz happy to jettison it but let's do it in another PR—don't wanna mix a hotfix with removing a feature :) one thing at a time

**Comment by iFrostizz** [comment]
Yep @Evalir , that was a blunder introduced in https://github.com/foundry-rs/foundry/pull/5692
We should probably write some tests that make sure that the debugger opens (no crash) with the right breakpoints so these would have been caught on CI later

**Comment by iFrostizz** [comment]
So just to make it clear is it okay if `forge debug` is currently broken ? Do we have a consensus to drop it ? (personally in favor of doing it)

**Comment by Evalir** [comment]
hmm, i think it requires a bit of discussion with the community as well as this is technically a breaking change—is it possible to fix up forge debug first? @iFrostizz

**Comment by mattsse** [comment]
can you elaborate on what is broken and why dropping forge debug is a good idea from your pov? @iFrostizz

**Comment by iFrostizz** [comment]
@Evalir good idea, let's fix this stuff first
@mattsse coming from https://github.com/foundry-rs/foundry/pull/4793#discussion_r1174305656 . `forge debug` is basically a slimmed down version of `forge script --debug` because it has much less options and hardcode stuffs just to satisfy the "script" part. It a big footgun imo, just like the scenario of the guy linked before. Because people may think that their script cannot be debugged while they should just use `forge script --debug`.
For legacy reasons, we could just copy the behavior of `forge script --debug` over `forge debug`, not sure if `clap` can flatten structs and hide fields from the command line conditionally cause a new struct can be created by composition not inheritance in Rust. dunno

**Comment by iFrostizz** [comment]
We can just flatten `DebugArgs` with `[clap(flatten)]` and make `ScriptArgs` composed from `DebugArgs`

**Comment by Evalir** [comment]
@iFrostizz i believe that should be good enough—will we include this in this PR?

**Review by Evalir — APPROVED** [review]
Thanks for the fix! Can you confirm you've QA'd this and works?

Also, unrelated, but I'd recommend enabling git commit signing—it's more secure and a good practice.

**Comment by iFrostizz** [comment]
@Evalir just dropped the stuff, let's see how you find it.
Sorry for the pgp stuff, just switched to a new DE, need to bring it back in a min

**Comment by iFrostizz** [comment]
- [x] forge test
  - [x] default
  - [x] debugger
  - [x] breakpoints
- [x] forge script
  - [x] default
  - [x] debugger
  - [x] breakpoints
- [x] forge debug
  - [x] default (debugger)
  - [x] breakpoints

look good!

**Comment by iFrostizz** [comment]
The last commit just makes `ScriptArgs` "inherits" (it's composed from) `DebugArgs` with the only diff that `debug` is hardcoded to `true` for `DebugArgs`. Keeping it for legacy purposes, but you can see how duplicated that command feels like. In short it's `forge script` without the `--debug` flag.

File: crates/forge/bin/cmd/debug.rs [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
 use foundry_common::evm::EvmArgs;
-use std::path::PathBuf;
 
 // Loads project's figment and merges the build cli arguments into it
 foundry_config::impl_figment_convert!(DebugArgs, opts, evm_opts);
 
 /// CLI arguments for `forge debug`.
-#[derive(Debug, Clone, Parser)]
+#[derive(Debug, Clone, Parser, Default)]
 pub struct DebugArgs {
```
Comment (Evalir) [inline crates/forge/bin/cmd/debug.rs:12]: hmm sorry for the bikeshed but slightly confused about this—wasn't the plan to `#[flatten]` `ScriptArgs`?

**Review by Evalir — CHANGES_REQUESTED** [review]

File: crates/forge/bin/cmd/debug.rs (hunk 1 above)
Comment (iFrostizz) [reply crates/forge/bin/cmd/debug.rs:12]: Do you mean Serde's flatten ? Isn't this only for Serialising / Deserializing ? I meant `[#clap(flatten)]` for it to be seamless to the cli

Comment (Evalir) [reply crates/forge/bin/cmd/debug.rs:12]: yup meant `#[clap(flatten)]`—i thought we were gonna do this instead of copying over all the script struct fields 😅

Comment (iFrostizz) [reply crates/forge/bin/cmd/debug.rs:12]: I didn't copied all the struct field, look at `ScriptArgs` 😁

Comment (Evalir) [reply crates/forge/bin/cmd/debug.rs:12]: @iFrostizz coming back to this, gotcha now i see. Frankly though this doesn't sit super well with me as we're now hiding a bunch of important args for script on `DebugArgs` and I'm not sure if this is good for flexibility. I think we can handle a bit of duplication for now—wdyt about duplicating the script args on debug? we'll be rewriting a few of these things over the coming months, so it's fine.

Comment (iFrostizz) [reply crates/forge/bin/cmd/debug.rs:12]: Okay, this won't be very future proof tho

**Comment by iFrostizz** [comment]
wait fuzzing is still broken for `forge debug`and `forge script --debug`

**Comment by iFrostizz** [comment]
@Evalir I get it, script's `--sig` flag is not supposed to be ran against any argument-based function, so no fuzzing for scripts, makes sense yeah ?
LGTM otherwise

**Comment by Evalir** [comment]
Yep, I think this makes sense—no reason why would you fuzz in a script!

thank you, will review soon!

**Comment by iFrostizz** [comment]
@mattsse friendly ping
Making sure you saw that review request

File: crates/forge/bin/cmd/debug.rs [hunk 2]
```diff
@@ -25,33 +26,137 @@ pub struct DebugArgs {
...
+    #[clap(long)]
+    pub broadcast: bool,
+
+    /// Skips on-chain simulation.
+    #[clap(long)]
+    pub skip_simulation: bool,
+
+    /// Relative percentage to multiply gas estimates by.
+    #[clap(long, short, default_value = "130")]
+    pub gas_estimate_multiplier: u64,
```
Comment (mattsse) [inline crates/forge/bin/cmd/debug.rs:62]: I'm having trouble understanding why changing all these fields is necessary in order to fix the referenced issue?

**Review by mattsse — CHANGES_REQUESTED** [review]
> I have introduced a bug that didn't filtered the tests correctly using the construct.

I want to do these always in two steps, fix first then refactor.

this is not really easy to review because unclear what the fix is and what refactor.

**Comment by mattsse** [comment]
> I have introduced a bug that didn't filtered the tests correctly using the construct.

this PR should then fix this bug,

could you please split this into two, because rn this pr does multiple things

**Comment by mattsse** [comment]
but maybe @Evalir has already checked it, in which case I'm okay with merging, codewise this looks fine, but the impl figment macro is now no longer accurate

**Comment by Evalir** [comment]
@mattsse im feeling uneasy about this PR due to the same reasons you mentioned—I'm in favor of splitting to make changes smaller and ideally also not screw up the figment macro

**Comment by Evalir** [comment]
Closing, this was fixed recently in another PR.
